parl/remote/future_mode/proxy_wrapper_nowait.py

```diff
@@ -0,0 +1,232 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import queue
+import time
+import threading
+
+from parl.remote.exceptions import FutureFunctionError
+from parl.remote.utils import RESERVED_NAME_ERROR_STR
+from parl.remote.future_mode.utils import FutureObject, CallingRequest
+
+
+def proxy_wrapper_nowait_func(remote_wrapper, max_memory):
+    '''
+    The 'proxy_wrapper_nowait_func' is defined on the top of class 'RemoteWrapper'
+    when using the `@remote_class(wait=False)` decorator. The main purpose is as follows:
+
+    1. set and get attributes of 'remoted_wrapper' and the corresponding 
+    remote models individually. (With 'proxy_wrapper_nowait_func', it is allowed to 
+    define a attribute (or method) of the same name in 'RemoteWrapper' and remote models.
+
+    2. provide asynchronous function calling. When user calling a function, he will get
+    the return(`FutureObject`) immediately and he need call the `get` function to get the
+    real return. For example:
+        ```python
+        import parl
+
+        @parl.remote_class(wait=False)
+        class Actor(object):
+            def __init__(self):
+                pass
+
+            def func(self):
+                # do something
+                return 0
+
+        parl.connect("localhost:8010")
+
+        actors = [Actor() for _ in range(10)]
+        future_objects = [actor.func() for actor in actors]
+        results = [future_obj.get() for future_obj in future_objects]
+        ```
+    '''
+
+    class ProxyWrapperNoWait(object):
+        def __init__(self, *args, **kwargs):
+            for reserved_name in [
+                    '__xparl_proxy_wrapper_nowait__', '__xparl_remote_class__',
+                    '__xparl_remote_class_max_memory__'
+            ]:
+                assert reserved_name not in kwargs, RESERVED_NAME_ERROR_STR.format(
+                    reserved_name)
+
+            kwargs['__xparl_proxy_wrapper_nowait__'] = self
+            kwargs['__xparl_remote_class__'] = remote_wrapper._original
+            kwargs['__xparl_remote_class_max_memory__'] = max_memory
+
+            self.xparl_remote_wrapper_calling_queue = queue.Queue()
```

Comment:
For all internal attributes of PARL, the names should start with `__xparl_`.

Reply:
done